autonomi/src/client/data/public.rs

```diff
@@ -44,100 +49,72 @@ impl Client {
+                debug!("Uploading {} chunks", chunks.len());
+
+                let mut failed_uploads = self
+                    .upload_chunks_with_retries(
+                        chunks
+                            .iter()
+                            .chain(std::iter::once(&data_map_chunk))
+                            .collect(),
+                        &receipt,
+                    )
+                    .await;
+
+                // Return the last chunk upload error
+                if let Some(last_chunk_fail) = failed_uploads.pop() {
+                    tracing::error!(
+                        "Error uploading chunk ({:?}): {:?}",
+                        last_chunk_fail.0.address(),
+                        last_chunk_fail.1
+                    );
+                    return Err(last_chunk_fail.1);
+                }
+
+                let record_count = chunks.len() + 1;
+
+                // Reporting
+                if let Some(channel) = self.client_event_sender.as_ref() {
+                    let tokens_spent = receipt
+                        .values()
+                        .map(|(_proof, price)| price.as_atto())
+                        .sum::<Amount>();
+
+                    let summary = UploadSummary {
+                        record_count,
+                        tokens_spent,
+                    };
+                    if let Err(err) = channel.send(ClientEvent::UploadComplete(summary)).await {
+                        error!("Failed to send client event: {err:?}");
+                    }
+                }
+
+                Ok(map_xor_name)
             }
-        }
-
-        Ok(map_xor_name)
-    }
-
-    /// Get a raw chunk from the network.
-    pub async fn chunk_get(&self, addr: ChunkAddr) -> Result<Chunk, GetError> {
-        info!("Getting chunk: {addr:?}");
-
-        let key = NetworkAddress::from_chunk_address(ChunkAddress::new(addr)).to_record_key();
-        debug!("Fetching chunk from network at: {key:?}");
-        let get_cfg = GetRecordCfg {
-            get_quorum: Quorum::One,
-            retry_strategy: None,
-            target_record: None,
-            expected_holders: HashSet::new(),
-            is_register: false,
-        };
-
-        let record = self
-            .network
-            .get_record_from_network(key, &get_cfg)
-            .await
-            .inspect_err(|err| error!("Error fetching chunk: {err:?}"))?;
-        let header = RecordHeader::from_record(&record)?;
-
-        if let RecordKind::Chunk = header.kind {
-            let chunk: Chunk = try_deserialize_record(&record)?;
-            Ok(chunk)
-        } else {
-            error!(
-                "Record kind mismatch: expected Chunk, got {:?}",
-                header.kind
-            );
-            Err(NetworkError::RecordKindMismatch(RecordKind::Chunk).into())
+            ClientMode::ReadOnly => Err(PutError::NoWallet),
```

Comment:
What does the mode protect us against? I'm wondering if it would be simpler without them?